main/src/ca/uwaterloo/flix/language/phase/Weeder.scala

```diff
@@ -880,49 +880,18 @@ object Weeder {
       }
 
     case ParsedAst.Expression.ApplicativeFor(sp1, frags, exp, sp2) =>
-      //
-      // Rewrites a ForA loop into a series of Applicative.ap calls:
-      //
-      //     forA (
-      //         x <- xs;
-      //         y <- ys
-      //     ) yield exp
-      //
-      // desugars to
-      //
-      //     Applicative.ap(Functor.map(x -> y -> exp, xs), ys)
-      //
-      val loc = mkSL(sp1, sp2).asSynthetic
-
-      if (frags.nonEmpty) {
-        val fqnAp = "Applicative.ap"
-        val fqnMap = "Functor.map"
-        val yieldExp = visitExp(exp, senv)
-
-        // Make lambda for Functor.map(lambda, ...). This lambda uses all patterns from the for-fragments.
-        val lambda = foldRight(frags)(yieldExp) {
-          case (ParsedAst.ForFragment.Generator(sp11, pat, _, sp12), acc) =>
-            mapN(visitPattern(pat)) {
-              case p => mkLambdaMatch(sp11, p, acc, sp12)
-            }
-        }
-
-        // Apply first fragment to Functor.map
-        val xs = visitExp(frags.head.exp, senv)
-        val baseExp = mapN(lambda, xs) {
-          case (l, x) => mkApplyFqn(fqnMap, List(l, x), loc)
-        }
-
-        // Apply rest of fragments to Applicative.ap
-        frags.tail.foldLeft(baseExp) {
-          case (acc, ParsedAst.ForFragment.Generator(sp11, _, fexp, sp12)) =>
-            mapN(acc, visitExp(fexp, senv)) {
-              case (a, e) => mkApplyFqn(fqnAp, List(a, e), mkSL(sp11, sp12).asSynthetic)
-            }
-        }
-      } else {
-        val err = WeederError.IllegalEmptyForFragment(loc)
-        WeededAst.Expr.Error(err).toSoftFailure(err)
+      val loc = mkSL(sp1, sp2)
+      val fs = mapN(traverse(frags)(visitForFragment(_, senv))) {
+        case fs1 =>
+          // This is ensured by the parser
+          fs1.collect { case t: WeededAst.ForFragment.Generator => t }
+      }
+      val e = visitExp(exp, senv)
+      flatMapN(fs, e) {
```

Comment:
I was wondering if this could be cleaned up? Like we have mapN, traverse, then flatMap etc.